### PR TITLE
Promotions

### PR DESCRIPTION
**Pull request by metalcupcake5**
- adds google form to the promotions
- allows "weighting" promotions to display some more than others
  - could probably be improved

image

### CODE REVIEW

File: src/lib.js [hunk 1]
```diff
@@ -512,6 +512,10 @@ async function processItems(base64, customTextures = false, packs, cacheOnly = f
       item.extra.skin = `PET_SKIN_${item.tag.ExtraAttributes.petInfo.skin}`;
     }
 
+    if (item.tag?.ExtraAttributes?.dye_item != undefined) {
+      item.extra.dye = item.tag.ExtraAttributes.dye_item;
```
Comment (dukio) [inline src/lib.js]: Ehm... wrong base branch?

**Comment by dukio** [comment]
About the randomWeighted function did you write or find it?

I was looking at https://dev.to/trekhleb/weighted-random-algorithm-in-javascript-1pdc and their implementation is slightly different... if you found it I'm just wondering if it's accurate (the fact that it's not using `const` where it should be or returning from the last `for` instead of breaking out of it is kinda worring me)

**Comment by metalcupcake5** [comment]
> did you write or find it

i found it, i do agree it could use some changing though. did some testing and it seems mostly accurate, i'll change a few things tho

**Review by dukio — CHANGES_REQUESTED** [review]

**Comment by dukio** [comment]
> > did you write or find it
> 
> i found it, i do agree it could use some changing though. did some testing and it seems mostly accurate, i'll change a few things tho

👍

File: src/lib.js (hunk 1 above)
Comment (metalcupcake5) [reply src/lib.js]: yikes mb

**Review by dukio — APPROVED** [review]

**Pull request merged**